### PR TITLE
add geckodriver 0.20.1 with rust

### PR DESCRIPTION
**Pull request by bollwyvl**
As suggested on #5759, actually building with `rust` is probably a better idea. Giving it a shot here...

### CODE REVIEW

**Comment by bollwyvl** [comment]
Hm, I don't see how turning up the verbosity and checking the versions made appveyor succeed, but I'll take it!

@mariusvniekerk

**Comment by mariusvniekerk** [comment]
@bollwyvl most of the ci's use lines of logs appearing as a measure of if the build has stalled.  If nothing write to the logs in a long while it may do something dumb 🤷‍♂️

File: recipes/geckodriver/build.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+set -ex
+rustc -V
+cargo -V
+
+if [ $(uname) = Darwin ] ; then
+    export CXXFLAGS="-arch $OSX_ARCH -stdlib=libc++ -std=c++11"
```
Comment (mariusvniekerk) [inline recipes/geckodriver/build.sh]: So you may want to ensure that you don't replace conda's `CXXFLAGS` and `CFLAGS`  here for compatibility.   
Does this build also invoke c / c++ somewhere internally?  If so you may want to add this to the meta.yaml

```yaml
  build:
    - {{ compiler('c') }}
    - {{ compiler('cxx') }}
```

**Comment by mariusvniekerk** [comment]
Care to take a look at this one @johanneskoester @pkgw? Not that certain about the *right way* to build rusty things with conda.

**Comment by pkgw** [comment]
Looks fine to me, modulo the comments already made!

**Comment by bollwyvl** [comment]
If nothing write to the logs in a long while it may do something dumb 🤷‍♂️
>
> Nah, Rusty is still pretty noisy, as it emits a line for each compiled
dependency. Total build time was well under the window when that kind of
whammy occurs.

Only happened in Windows, and I haven't gotten my local vm up with vs2015
which had link.exe so I haven't been able to reproduce, or look at the
generated build script for the regex package (where the error was reported).

>
>

**Comment by bollwyvl** [comment]
The flags were cargo (ha!) culted from tectonic-feedstock... But it had
lots of other c-based deps. I'll de-flag and see... Works on Windows on
appveyor, but that proves very little.

> *@mariusvniekerk* commented on this pull request.
>
> Does this build also invoke c / c++ somewhere internally?
>

**Comment by bollwyvl** [comment]
Okiedoke, appears to build just fine without the flags, but it looks like it is linking against system libgcc. So i guess we in fact want the compilers?

Here's the inspect output:

circle:
```
+ conda inspect linkages -p /home/conda/feedstock_root/build_artifacts/geckodriver_1525473056572/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh geckodriver
WARNING: pyldd disagrees with ldd/otool. This will not cause any
WARNING: problems for this build, but please file a bug at:
WARNING: https://github.com/conda/conda-build
WARNING: and (if possible) attach file /home/conda/feedstock_root/build_artifacts/geckodriver_1525473056572/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/bin/geckodriver
WARNING: 
ldd/otool gives:
('linux-vdso.so.1', '')
('libdl.so.2', '/lib64/libdl.so.2')
('librt.so.1', '/lib64/librt.so.1')
('libpthread.so.0', '/lib64/libpthread.so.0')
('libgcc_s.so.1', '/lib64/libgcc_s.so.1')
('libc.so.6', '/lib64/libc.so.6')
pyldd gives:
('libdl.so.2', '/lib64/libdl.so.2')
('librt.so.1', '/lib64/librt.so.1')
('ld-linux-x86-64.so.2', '/lib64/ld-linux-x86-64.so.2')
('libc.so.6', '/lib64/libc.so.6')
('libpthread.so.0', '/lib64/libpthread.so.0')
('libgcc_s.so.1', '/lib64/libgcc_s.so.1')

Diffs
{('linux-vdso.so.1', '')}
Diffs
{('ld-linux-x86-64.so.2', '/lib64/ld-linux-x86-64.so.2')}
geckodriver
-----------

system:
    libc.so.6 (/lib64/libc.so.6)
    libdl.so.2 (/lib64/libdl.so.2)
    libgcc_s.so.1 (/lib64/libgcc_s.so.1)
    libpthread.so.0 (/lib64/libpthread.so.0)
    librt.so.1 (/lib64/librt.so.1)
    linux-vdso.so.1 ()

not found:


+ exit 0
```

and osx:
```
+conda inspect linkages -p /Users/travis/miniconda/conda-bld/geckodriver_1525473199034/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place geckodriver
geckodriver
-----------
system:
    libSystem.B.dylib (/usr/lib/libSystem.B.dylib)
    libresolv.9.dylib (/usr/lib/libresolv.9.dylib)
not found:
+conda inspect objects -p /Users/travis/miniconda/conda-bld/geckodriver_1525473199034/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place geckodriver
geckodriver
-----------
bin/geckodriver
  filetype: EXECUTE
  rpath: /Users/travis/miniconda/conda-bld/geckodriver_1525473199034/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib:@loader_path/../lib/
+exit 0
```

**Comment by bollwyvl** [comment]
More [appveyor fun](https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.23891#L917):

```
= note: libbzip2-d0af07eec02943bd.rlib(bzip2-d0af07eec02943bd.bzip21-ce1dfab364a2987ecd2b56cf1ff4036f.rs.rcgu.o) : error LNK2019: unresolved external symbol BZ2_bzCompressInit referenced in function _ZN5bzip23mem8Compress3new17hfa91210b740d3e98E
          libbzip2-d0af07eec02943bd.rlib(bzip2-d0af07eec02943bd.bzip21-ce1dfab364a2987ecd2b56cf1ff4036f.rs.rcgu.o) : error LNK2019: unresolved external symbol BZ2_bzCompress referenced in function _ZN5bzip23mem8Compress8compress17h6129e40505c9ab65E
          libbzip2-d0af07eec02943bd.rlib(bzip2-d0af07eec02943bd.bzip21-ce1dfab364a2987ecd2b56cf1ff4036f.rs.rcgu.o) : error LNK2019: unresolved external symbol BZ2_bzDecompressInit referenced in function _ZN5bzip23mem10Decompress3new17hdc77fdef3d3aaa65E
          libbzip2-d0af07eec02943bd.rlib(bzip2-d0af07eec02943bd.bzip21-ce1dfab364a2987ecd2b56cf1ff4036f.rs.rcgu.o) : error LNK2019: unresolved external symbol BZ2_bzDecompress referenced in function _ZN5bzip23mem10Decompress10decompress17hed6a695bda98eec1E
          libbzip2-d0af07eec02943bd.rlib(bzip2-d0af07eec02943bd.bzip21-ce1dfab364a2987ecd2b56cf1ff4036f.rs.rcgu.o) : error LNK2019: unresolved external symbol BZ2_bzCompressEnd referenced in function _ZN65_$LT$bzip2..mem..DirCompress$u20$as$u20$bzip2..mem..Direction$GT$7destroy17h52342ee9843ed186E
          libbzip2-d0af07eec02943bd.rlib(bzip2-d0af07eec02943bd.bzip21-ce1dfab364a2987ecd2b56cf1ff4036f.rs.rcgu.o) : error LNK2019: unresolved external symbol BZ2_bzDecompressEnd referenced in function _ZN67_$LT$bzip2..mem..DirDecompress$u20$as$u20$bzip2..mem..Direction$GT$7destroy17h910ad67806095456E
          C:\bld\geckodriver_1525477156401\work\target\release\deps\geckodriver-fdd61103708781cc.exe : fatal error LNK1120: 6 unresolved externals
          
error: aborting due to previous error
error: Could not compile `geckodriver`.
Caused by:
  process didn't exit successfully: `rustc --crate-name geckodriver src\main.rs --crate-type bin --emit=dep-info,link -C opt-level=3 -C metadata=fdd61103708781cc -C extra-filename=-fdd61103708781cc --out-dir C:\bld\geckodriver_1525477156401\work\target\release\deps -L dependency=C:\bld\geckodriver_1525477156401\work\target\release\deps --extern webdriver=C:\bld\geckodriver_1525477156401\work\target\release\deps\libwebdriver-04c9095e4426a393.rlib --extern mozrunner=C:\bld\geckodriver_1525477156401\work\target\release\deps\libmozrunner-4afbc6cfde442179.rlib --extern mozversion=C:\bld\geckodriver_1525477156401\work\target\release\deps\libmozversion-1d317c309f84f4d1.rlib --extern chrono=C:\bld\geckodriver_1525477156401\work\target\release\deps\libchrono-c377078e70725b52.rlib --extern log=C:\bld\geckodriver_1525477156401\work\target\release\deps\liblog-3bfbe0c96567f897.rlib --extern lazy_static=C:\bld\geckodriver_1525477156401\work\target\release\deps\liblazy_static-b4370319aaa2e4d2.rlib --extern rustc_serialize=C:\bld\geckodriver_1525477156401\work\target\release\deps\librustc_serialize-7a85f1ce7d31075b.rlib --extern clap=C:\bld\geckodriver_1525477156401\work\target\release\deps\libclap-b8ae48eace76701c.rlib --extern regex=C:\bld\geckodriver_1525477156401\work\target\release\deps\libregex-a0d8a8debb9203e2.rlib --extern hyper=C:\bld\geckodriver_1525477156401\work\target\release\deps\libhyper-38f72d504ad6fdd1.rlib --extern zip=C:\bld\geckodriver_1525477156401\work\target\release\deps\libzip-89b94f1003375770.rlib --extern uuid=C:\bld\geckodriver_1525477156401\work\target\release\deps\libuuid-323cfa8e7c8bd872.rlib --extern mozprofile=C:\bld\geckodriver_1525477156401\work\target\release\deps\libmozprofile-bf8a57cf919e12f0.rlib -L native=C:\bld\geckodriver_1525477156401\work\target\release\build\bzip2-sys-6ea84d67eea82f61\out` (exit code: 101)
```

Seems odd, but trying with `bzip2`.

**Comment by bollwyvl** [comment]
Once again, we are green.... but I don't know why, as even using the legacy `msvc` flag it [successfully built](https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.23897#L187) on both of the compilers:
```
Resolved dependencies, will be built in the following order:
    geckodriver-0.20.1-cxx_compiler_vs2008-c_compiler_vs2008-on-win-64
    geckodriver-0.20.1-cxx_compiler_vs2015-c_compiler_vs2015-on-win-64
```
...5 minutes later...
```
TEST END: C:\bld\win-64\geckodriver-0.20.1-hc56fc5f_0.tar.bz2
BUILD START: ['geckodriver-0.20.1-h6538335_0.tar.bz2']
```
...5 minutes later...
```
TEST END: C:\bld\win-64\geckodriver-0.20.1-h6538335_0.tar.bz2
# Automatic uploading is disabled
# If you want to upload package(s) to anaconda.org later, type:
anaconda upload C:\bld\win-64\geckodriver-0.20.1-hc56fc5f_0.tar.bz2
anaconda upload C:\bld\win-64\geckodriver-0.20.1-h6538335_0.tar.bz2

```

**Comment by bollwyvl** [comment]
Here's a new one:
```
          LINK : warning LNK4044: unrecognized option '/NATVIS:C:\bld\geckodriver_1525489758022\_h_env\Library\lib\rustlib\etc\liballoc.natvis'; ignored
          LINK : warning LNK4044: unrecognized option '/NATVIS:C:\bld\geckodriver_1525489758022\_h_env\Library\lib\rustlib\etc\libcore.natvis'; ignored
          libbzip2_sys-f6dafbd0534fd078.rlib(compress.o) : error LNK2019: unresolved external symbol __report_rangecheckfailure referenced in function sendMTFValues
          libbzip2_sys-f6dafbd0534fd078.rlib(blocksort.o) : error LNK2001: unresolved external symbol __report_rangecheckfailure
          C:\bld\geckodriver_1525489758022\work\target\release\deps\geckodriver-fdd61103708781cc.exe : fatal error LNK1120: 1 unresolved externals
```

**Comment by bollwyvl** [comment]
Would definitely appreciate some assist on this: how might I force only building against vs2015 on windows? That path does seem to flake at times, but seems to work more reliably.

**Comment by bollwyvl** [comment]
Trying to force `vs2015` on windows with python 3.6...

File: recipes/geckodriver/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+source:
+  fn: {{ name }}-{{ version }}.tar.gz
+
+  url: https://github.com/mozilla/{{ name }}/archive/v{{ version }}.tar.gz
+  sha256: 0bee8be4841147528c41417e4eb1f44eaddd7aa16b267d6237ec2abafecf71b2
+
+build:
+  number: 0
+  skip: true  # [win and vc!=14]
```
Comment (bollwyvl) [inline recipes/geckodriver/meta.yaml]: c/f https://github.com/conda-forge/ninja-feedstock/pull/7/files#diff-e178b687b10a71a3348107ae3154e44cR14

**Comment by bollwyvl** [comment]
Ok, so reverting to `msvc_compiler: 14.0  # [win]` gets us back to green.

I am wondering if we don't need to build `rust` and `cargo` with the new compilers in order to build downstream stuff. Probably out of scope for this little package, but hey, at least it's not self-hosting or something.

Anyhow, I'd really like to get this out the door! Firefox 60.0 has landed, and no doubt there will be something that demands a new release of geckodriver, but selfishly i have stuff for older foxen that would like this version!

**Comment by bgruening** [comment]
Thanks @bollwyvl!

**Comment by jakirkham** [comment]
FYI we are having conversion problems as noted in issue ( https://github.com/conda-forge/staged-recipes/issues/5846 ).

**Pull request merged**